File: Leetcode/82-Remove-Duplicates-from-Sorted-List-II/RemoveDuplicatesFromSortedListII.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func deleteDuplicates(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	dummy := new(ListNode)
	dummy.Next = head
	cur := dummy
	isDup := false
	for cur != nil && cur.Next != nil {
		if cur.Next.Next != nil && cur.Next.Val == cur.Next.Next.Val {
			isDup = true
			cur.Next.Next = cur.Next.Next.Next
		} else {
			if isDup {
				isDup = false
				cur.Next = cur.Next.Next
			} else {
				cur = cur.Next
			}
		}
	}
	return dummy.Next
}

File: Leetcode/82-Remove-Duplicates-from-Sorted-List-II/RemoveDuplicatesFromSortedListII.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head is None or head.next is None:
            return head
        dummy = ListNode(0)
        dummy.next = head
        cur = dummy
        isDup = False
        while cur is not None and cur.next is not None:
            if cur.next.next is not None and cur.next.val == cur.next.next.val:
                isDup = True
                cur.next.next = cur.next.next.next
            else :
                if isDup:
                    isDup = False
                    cur.next = cur.next.next
                else:
                    cur = cur.next
        return dummy.next
    
